src/Castle.Core/DynamicProxy/ProxyGenerationOptions.cs

```diff
@@ -204,7 +208,53 @@ public override int GetHashCode()
 			result = 29*result + (Selector != null ? 1 : 0);
 			result = 29*result + MixinData.GetHashCode();
 			result = 29*result + (BaseTypeForInterfaceProxy != null ? BaseTypeForInterfaceProxy.GetHashCode() : 0);
+		    result = 29*result + GetContentsHashCode(AdditionalAttributes);
 			return result;
 		}
+
+        private static int GetContentsHashCode<T>(IList<T> list)
+        {
+            var result = 0;
+            for (var i = 0; i < list.Count; i++)
+            {
+                if (list[i] != null)
+                {
+                    // simply XOR since order does not matter
+                    result ^= list[i].GetHashCode();
+                }
+            }
+
+            return result;
+        }
+
+        private static bool ListsEquivalent<T>(IList<T> listA, IList<T> listB)
+        {
+            if (listA.Count != listB.Count)
+            {
+                return false;
+            }
+
+            // order is not important, just make sure that each entry in A is also found in B

```

Comment:
An option then would be to copy the destination items to a local array and set the value to `null` once it has already been "found" with another. This would force `Equals` to have the same number of "same" objects in both lists. That as well as using `add` instead of `xor` in `GetHashCode` should satisfy the cases that you have brought up. I'll try to get the changes in later today.